lib/src/modules/task.cc

```diff
@@ -426,9 +438,110 @@ static fs::path getCachedTaskFile(const fs::path& task_cache_dir,
+        }
+    }
+
+    return install_dir;
+}
+
+// Get a unique list of "lib" files that the task has specified in metadata
+// this involves enumerating files when directories are requested
+std::set<std::string> getMultiFiles(std::vector<std::string> const& meta_files, std::vector<std::string>  const& impl_files, std::vector<lth_jc::JsonContainer> const& files) {
+    // build set of unique files to download and exclude the selected task executable
+    std::set<std::string> download_set(meta_files.begin(), meta_files.end());
+    download_set.insert(impl_files.begin(), impl_files.end());
+    std::vector<std::string> directories;
+
+    // expand files from directories specified by trailing '/'
+    for (auto f_name : download_set) {
+        if (f_name.back() == '/') {
+            directories.push_back(f_name);
+            for (auto f : files) {
+                if (f.get<std::string>("filename").substr(0, f_name.size()) == f_name) {
+                    download_set.insert(f.get<std::string>("filename"));
+                }
+            }
+        }
+    }
+
+    // remove directories from set
+    for (auto f_name : directories) {
+      download_set.erase(f_name);
+    }
+
+    return download_set;
+}
+
+// iterate over unique set of filenames to support task
+// copy file from cache to install_dir
+fs::path Task::downloadMultiFile(std::vector<lth_jc::JsonContainer> const& files, std::set<std::string> const& download_set) {
+    auto install_dir =  createInstallDir(task_cache_dir_, download_set);
+    for (auto file_name : download_set) {
+        // get file object info based on name
+        auto file_object = selectLibFile(files, file_name);
+
+        // get file from cache, download if necessary
+        auto lib_file = getCachedTaskFile(task_cache_dir_,
+                                       task_cache_dir_mutex_,
+                                       master_uris_,
+                                       task_download_connect_timeout_,
+                                       task_download_timeout_,
+                                       client_,
+                                       file_object);
+
+        // copy to expected location in install_dir
+        auto t = install_dir / fs::path(file_name);
+        fs::copy_file(lib_file, install_dir / fs::path(file_name));
+    }
+
+    return install_dir;
+}
+
+lth_jc::JsonContainer Task::selectLibFile(std::vector<lth_jc::JsonContainer> const& files,
+                                            std::string const& file_name)
+{
+    // Select file based on impl.
+    auto file = std::find_if(files.cbegin(), files.cend(),
+        [&](lth_jc::JsonContainer const& file) {
+            auto filename = file.get<std::string>("filename");
+            return filename == file_name;
+        });
+
+    if (file == files.cend()) {
+        throw Module::ProcessingError {
+            lth_loc::format("'{1}' file requested by implementation not found", file_name) };
```

Comment:
This error message doesn't seem very clear either. It can work, but implementation isn't obviously related in my mind.

Reply:
Agreed, this is unclear. Updated.